{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a1b2c3d4",
   "metadata": {},
   "source": [
    "# Multiple Sclerosis Classifier\n",
    "\n",
    "This notebook contains a simple classifier for multiple sclerosis (MS) based on gene expression data. The goal of this project is to learn how to write a simple classifier for healthy and unhealthy patients.\n",
    "\n",
    "The data is from a study comparing gene expression in peripheral blood samples between healthy human individuals and patients with relapsing-remitting multiple sclerosis (RRMS). The data was generated using RNA sequencing (RNA-Seq) to identify differences in mRNA levels between the two groups.\n",
    "\n",
    "If you prefer you can use alternative datasets, classifying different diseases. Data can be acquired [here](https://www.ebi.ac.uk/gxa/experiments?experimentType=%22Differential%22&species=homo+sapiens). \n",
    "Choose datasets with at least two experimental factors. \n",
    "The default dataset (MS) was from [this](https://www.ebi.ac.uk/gxa/experiments/E-GEOD-66573/Results) link.\n",
    "In case you are unsure how to handle different datasets, an alternative [dataset](https://www.ebi.ac.uk/gxa/experiments/E-ENAD-46/Results) is included as an example and can be loaded by changing `DATASET = lung` further below.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c3d4e5",
   "metadata": {},
   "source": [
    "## Data Files\n",
    "\n",
    "The data is located in the `data/` folder and consists of two datasets: Multiple Sclerosis (MS) and Lung.\n",
    "\n",
    "### Multiple Sclerosis (MS)\n",
    "\n",
    "The MS data is located in the `data/ms/` folder and consists of the following files:\n",
    "\n",
    "*   `data/ms/raw-counts.tsv`: This file contains the raw gene counts for each sample. The rows represent genes and the columns represent samples.\n",
    "*   `data/ms/experiment-design.tsv`: This file describes the samples, indicating which ones are from healthy controls and which are from RRMS patients, along with other relevant metadata.\n",
    "*   `data/ms/query-results.tsv`: This file contains the query results from the Expression Atlas database.\n",
    "\n",
    "### Lung\n",
    "\n",
    "The Lung data is located in the `data/lung/` folder and consists of the following files:\n",
    "\n",
    "*   `data/lung/raw-counts.tsv`: This file contains the raw gene counts for each sample. The rows represent genes and the columns represent samples.\n",
    "*   `data/lung/experiment-design.tsv`: This file describes the samples and their metadata.\n",
    "*   `data/lung/query-results.tsv`: This file contains the query results from the Expression Atlas database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c3d4e6",
   "metadata": {},
   "source": [
    "### Your Tasks\n",
    "\n",
    "Your main task is to improve the classifier by selecting better features. This involves the following steps:\n",
    "\n",
    "1.  **Familiarize yourself with the covariance matrix:** The covariance matrix is a powerful tool for understanding the relationships between features. The current version of this notebook generates a covariance matrix for the first 5 genes.\n",
    "2.  **Normalization:** The data is normalized before calculating the covariance matrix. Why is this important? What would happen if you didn't normalize the data?\n",
    "3.  **Feature Selection:** The current version of the notebook only uses the first 5 genes as features. You should experiment with different sets of features to see if you can improve the accuracy of the classifier. You can use the information from the `query results` dataframe to select genes that are differentially expressed between the two groups.\n",
    "4.  **Interpret the covariance matrix:** What does it mean if a value in the covariance matrix is close to 0? What does it mean if it is close to 1? How can you use this information to select features?\n",
    "5.  **Implement k-fold cross validation:** Since we have limited data (only 14 samples total), a single train/test split may not give us a reliable estimate of model performance. Implement k-fold cross validation to get a more robust evaluation of your classifier. Consider what value of k would be appropriate given our small dataset size. Compare the results from k-fold cross validation with the simple train/test split approach currently used in the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f05e4f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure that packages are installed\n",
    "!pip install pandas scikit-learn matplotlib numpy\n",
    "\n",
    "# Import necessary libraries\n",
    "import pandas as pd\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.metrics import accuracy_score"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d4e5f6",
   "metadata": {},
   "source": [
    "### Raw Counts\n",
    "\n",
    "This file contains the raw gene counts for each sample. The rows represent genes and the columns represent samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "new_markdown_cell",
   "metadata": {},
   "source": [
    "### Select the dataset to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "new_code_cell",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET = \"ms\" # or \"lung\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b18b20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = f\"data/{DATASET}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53869909",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_raw_counts = pd.read_csv(f\"{DATA_DIR}/raw-counts.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_raw_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5f6g7h8",
   "metadata": {},
   "source": [
    "### Experiment Design\n",
    "\n",
    "This file describes the samples, indicating which ones are from healthy controls and which are from RRMS patients, along with other relevant metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994cdde5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_experiment_design = pd.read_csv(f\"{DATA_DIR}/experiment-design.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_experiment_design"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6g7h8i9",
   "metadata": {},
   "source": [
    "### Query Results\n",
    "\n",
    "This file contains the query results from the Expression Atlas database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac40d284",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_query_results = pd.read_csv(f\"{DATA_DIR}/query-results.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_query_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "g7h8i9j0",
   "metadata": {},
   "source": [
    "### Data Preparation\n",
    "\n",
    "Before we can train a classifier, we need to prepare the data. This involves selecting features and labels from the dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49036aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_features_of_patient(df, patient_id):\n",
    "    \"\"\"\n",
    "    Get the raw gene counts of a specific patient from the DataFrame.\n",
    "    \n",
    "    Parameters:\n",
    "    df (pd.DataFrame): DataFrame containing patient data.\n",
    "    patient_id (str): ID of the patient to retrieve features for. You can find the IDs in the first column of the analytics dataframe.\n",
    "    \n",
    "    Returns:\n",
    "    pd.Series: Features of the specified patient.\n",
    "    \"\"\"\n",
    "    return df[patient_id][:5].to_list() # TODO: Select features smartly, not just the first 5 (genes)\n",
    "\n",
    "get_features_of_patient(df_raw_counts, \"SRR1839791\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "251a1ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dataset(df_raw_counts, df_experiment_design): \n",
    "    \"\"\"\n",
    "    Get the entire dataset of features and labels.\n",
    "    \n",
    "    Parameters:\n",
    "    df_raw_counts (pd.DataFrame): DataFrame containing raw gene counts.\n",
    "    df_experiment_design (pd.DataFrame): DataFrame containing experiment design data.\n",
    "    \n",
    "    Returns:\n",
    "    tuple: Features and labels for the dataset.\n",
    "    \"\"\"\n",
    "    patient_ids = df_experiment_design.index.to_list()\n",
    "    features = [get_features_of_patient(df_raw_counts, patient_id) for patient_id in patient_ids]\n",
    "    labels = df_experiment_design[\"Sample Characteristic[disease]\"].to_list()\n",
    "    unique_labels = list(set(labels))\n",
    "    labels = [unique_labels.index(label) for label in labels] # Convert labels to numbers\n",
    "    return features, labels\n",
    "\n",
    "features, labels = get_dataset(df_raw_counts, df_experiment_design)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "h8i9j0k1",
   "metadata": {},
   "source": [
    "### Train/Test Split\n",
    "\n",
    "We split the data into a training set and a testing set. The training set is used to train the classifier, and the testing set is used to evaluate its performance. The main bottleneck we have with that project is the lack of enough data. In fact, we have 14 cases in total with 8 and 6 cases respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "864af4ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_features, test_features, train_labels, test_labels = train_test_split(features, labels, test_size=0.5, random_state=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "i9j0k1l2",
   "metadata": {},
   "source": [
    "### Train the Classifier\n",
    "\n",
    "We use a simple Decision Tree Classifier to train the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb1759f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "classifier = DecisionTreeClassifier(random_state=2)\n",
    "classifier = classifier.fit(train_features, train_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "j0k1l2m3",
   "metadata": {},
   "source": [
    "### Evaluate the Classifier\n",
    "\n",
    "We use the accuracy score to evaluate the performance of the classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a18a6cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "predicted_test_labels = classifier.predict(test_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7498d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "accuracy = accuracy_score(test_labels, predicted_test_labels)\n",
    "\n",
    "print(f\"Baseline accuracy: {accuracy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "covariance_matrix_cell",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "# Load the data\n",
    "data = np.array(features)\n",
    "\n",
    "# Transpose the data so that rows are samples and columns are features (genes)\n",
    "data = data\n",
    "\n",
    "# Normalize the data\n",
    "scaler = StandardScaler()\n",
    "data_normalized = scaler.fit_transform(data)\n",
    "\n",
    "# Calculate the covariance matrix\n",
    "covariance_matrix = np.cov(data_normalized, rowvar=False)\n",
    "\n",
    "# Plot the covariance matrix\n",
    "plt.figure(figsize=(10, 8))\n",
    "plt.imshow(covariance_matrix, cmap='viridis', interpolation='nearest')\n",
    "plt.title('Covariance Matrix of Features')\n",
    "plt.colorbar()\n",
    "plt.xlabel('Features')\n",
    "plt.ylabel('Features')\n",
    "plt.savefig('covariance_matrix.png')\n",
    "print(\"Covariance matrix plot saved to covariance_matrix.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0971fce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_single_gene_feature(df_raw_counts, df_experiment_design, gene): \n",
    "    \"\"\"\n",
    "    Get the single gene feature for all patients in the dataset.\n",
    "    \n",
    "    Parameters:\n",
    "    df_raw_counts (pd.DataFrame): DataFrame containing raw gene counts.\n",
    "    df_experiment_design (pd.DataFrame): DataFrame containing experiment design data.\n",
    "    \n",
    "    Returns:\n",
    "    tuple: Features and labels for the dataset.\n",
    "    \"\"\"\n",
    "    patient_ids = df_experiment_design.index.to_list()\n",
    "\n",
    "    gene_index = df_raw_counts.index.get_loc(gene)\n",
    "\n",
    "    features = [df_raw_counts[patient_id][gene_index:gene_index + 1].to_list() for patient_id in patient_ids]\n",
    "    labels = df_experiment_design[\"Sample Characteristic[disease]\"].to_list()\n",
    "    unique_labels = list(set(labels))\n",
    "    labels = [unique_labels.index(label) for label in labels] # Convert labels to numbers\n",
    "    return features, labels\n",
    "\n",
    "single_feature, labels = get_single_gene_feature(df_raw_counts, df_experiment_design, \"ENSG00000250696\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a8a6a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "if DATASET == \"ms\":\n",
    "    positive_label = \"relapsing-remitting multiple sclerosis\"\n",
    "    negative_label = \"normal\"\n",
    "else: \n",
    "    positive_label = \"NHBE\"\n",
    "    negative_label = \"A549\"\n",
    "\n",
    "X = np.array(single_feature)[:, 0]\n",
    "y = np.array(labels)\n",
    "\n",
    "pos = X[y == 1]\n",
    "neg = X[y == 0]\n",
    "\n",
    "plt.plot(pos, np.zeros_like(pos), 'x')\n",
    "plt.plot(neg, np.zeros_like(neg), 'o')\n",
    "plt.show()\n",
    "\n",
    "#TODO: Add 2D plots as well"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
