{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a1b2c3d4",
   "metadata": {},
   "source": [
    "# Multiple Sclerosis Classifier\n",
    "\n",
    "This notebook contains a simple classifier for multiple sclerosis (MS) based on gene expression data. The goal of this project is to learn how to write a simple classifier for healthy and unhealthy patients.\n",
    "\n",
    "The data is from a study comparing gene expression in peripheral blood samples between healthy human individuals and patients with relapsing-remitting multiple sclerosis (RRMS). The data was generated using RNA sequencing (RNA-Seq) to identify differences in mRNA levels between the two groups.\n",
    "\n",
    "If you prefer you can use alternative datasets, classifying different diseases. Data can be acquired [here](https://www.ebi.ac.uk/gxa/experiments?experimentType=%22Differential%22&species=homo+sapiens). \n",
    "Choose datasets with at least two experimental factors. \n",
    "The default dataset (MS) was from [this](https://www.ebi.ac.uk/gxa/experiments/E-GEOD-66573/Results) link.\n",
    "In case you are unsure how to handle different datasets, an alternative [dataset](https://www.ebi.ac.uk/gxa/experiments/E-ENAD-46/Results) is included as an example and can be loaded by changing `DATASET = lung` further below.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c3d4e5",
   "metadata": {},
   "source": [
    "## Data Files\n",
    "\n",
    "The data is located in the `data/` folder and consists of two datasets: Multiple Sclerosis (MS) and Lung.\n",
    "\n",
    "### Multiple Sclerosis (MS)\n",
    "\n",
    "The MS data is located in the `data/ms/` folder and consists of the following files:\n",
    "\n",
    "*   `data/ms/raw-counts.tsv`: This file contains the raw gene counts for each sample. The rows represent genes and the columns represent samples.\n",
    "*   `data/ms/experiment-design.tsv`: This file describes the samples, indicating which ones are from healthy controls and which are from RRMS patients, along with other relevant metadata.\n",
    "*   `data/ms/query-results.tsv`: This file contains the query results from the Expression Atlas database.\n",
    "\n",
    "### Lung\n",
    "\n",
    "The Lung data is located in the `data/lung/` folder and consists of the following files:\n",
    "\n",
    "*   `data/lung/raw-counts.tsv`: This file contains the raw gene counts for each sample. The rows represent genes and the columns represent samples.\n",
    "*   `data/lung/experiment-design.tsv`: This file describes the samples and their metadata.\n",
    "*   `data/lung/query-results.tsv`: This file contains the query results from the Expression Atlas database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c3d4e6",
   "metadata": {},
   "source": [
    "### Your Tasks\n",
    "\n",
    "Your main task is to improve the classifier by selecting better features. This involves the following steps:\n",
    "\n",
    "1.  **Familiarize yourself with the covariance matrix:** The covariance matrix is a powerful tool for understanding the relationships between features. The current version of this notebook generates a covariance matrix for the first 5 genes.\n",
    "2.  **Normalization:** The data is normalized before calculating the covariance matrix. Why is this important? What would happen if you didn't normalize the data?\n",
    "3.  **Feature Selection:** The current version of the notebook only uses the first 5 genes as features. You should experiment with different sets of features to see if you can improve the accuracy of the classifier. You can use the information from the `query results` dataframe to select genes that are differentially expressed between the two groups.\n",
    "4.  **Interpret the covariance matrix:** What does it mean if a value in the covariance matrix is close to 0? What does it mean if it is close to 1? How can you use this information to select features?\n",
    "5.  **Implement k-fold cross validation:** Since we have limited data (only 14 samples total), a single train/test split may not give us a reliable estimate of model performance. Implement k-fold cross validation to get a more robust evaluation of your classifier. Consider what value of k would be appropriate given our small dataset size. Compare the results from k-fold cross validation with the simple train/test split approach currently used in the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "4f05e4f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (2.3.3)\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (1.7.2)\n",
      "Requirement already satisfied: matplotlib in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (3.10.7)\n",
      "Requirement already satisfied: numpy in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (1.26.4)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from pandas) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from pandas) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from pandas) (2025.2)\n",
      "Requirement already satisfied: scipy>=1.8.0 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from scikit-learn) (1.15.3)\n",
      "Requirement already satisfied: joblib>=1.2.0 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from scikit-learn) (1.5.2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from scikit-learn) (3.6.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (4.60.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (1.4.9)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (25.0)\n",
      "Requirement already satisfied: pillow>=8 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (11.2.1)\n",
      "Requirement already satisfied: pyparsing>=3 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from matplotlib) (3.2.5)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\theo\\appdata\\local\\schrodinger\\pymol2\\lib\\site-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n"
     ]
    }
   ],
   "source": [
    "# Ensure that packages are installed\n",
    "!pip install pandas scikit-learn matplotlib numpy\n",
    "\n",
    "# Import necessary libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.feature_selection import VarianceThreshold, SelectKBest, f_classif\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import KFold"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d4e5f6",
   "metadata": {},
   "source": [
    "### Raw Counts\n",
    "\n",
    "This file contains the raw gene counts for each sample. The rows represent genes and the columns represent samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "new_markdown_cell",
   "metadata": {},
   "source": [
    "### Select the dataset to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "new_code_cell",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET = \"data\" # or \"lung\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "0b18b20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = DATASET  #f\"data/{DATASET}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "53869909",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original number of genes: 58735\n",
      "Filtered number of genes: 33811\n",
      "Removed: 24924 genes with no expression\n"
     ]
    }
   ],
   "source": [
    "# Remove all genes with 0 expression in all 14 samples\n",
    "\n",
    "df_raw_counts = pd.read_csv(f\"{DATA_DIR}/raw-counts.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_raw_counts\n",
    "df_raw_counts = df_raw_counts.apply(pd.to_numeric, errors='coerce')\n",
    "df_filtered = df_raw_counts.loc[df_raw_counts.nunique(axis=1) > 1]\n",
    "#df_filtered = df_filtered[(df_filtered > 5).any(axis=1)]\n",
    "print(\"Original number of genes:\", df_raw_counts.shape[0])\n",
    "print(\"Filtered number of genes:\", df_filtered.shape[0])\n",
    "print(\"Removed:\", df_raw_counts.shape[0] - df_filtered.shape[0], \"genes with no expression\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5f6g7h8",
   "metadata": {},
   "source": [
    "### Experiment Design\n",
    "\n",
    "This file describes the samples, indicating which ones are from healthy controls and which are from RRMS patients, along with other relevant metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "994cdde5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Characteristic[disease]</th>\n",
       "      <th>Sample Characteristic Ontology Term[disease]</th>\n",
       "      <th>Sample Characteristic[ethnic group]</th>\n",
       "      <th>Sample Characteristic Ontology Term[ethnic group]</th>\n",
       "      <th>Sample Characteristic[organism]</th>\n",
       "      <th>Sample Characteristic Ontology Term[organism]</th>\n",
       "      <th>Sample Characteristic[organism part]</th>\n",
       "      <th>Sample Characteristic Ontology Term[organism part]</th>\n",
       "      <th>Factor Value[disease]</th>\n",
       "      <th>Factor Value Ontology Term[disease]</th>\n",
       "      <th>Analysed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1839791</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839792</th>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839793</th>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839794</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839795</th>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839796</th>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839797</th>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839798</th>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839799</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839800</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839801</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839802</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839803</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1839804</th>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HANCESTRO_0005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>blood</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000178</td>\n",
       "      <td>normal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Sample Characteristic[disease]  \\\n",
       "Run                                                  \n",
       "SRR1839791                                  normal   \n",
       "SRR1839792  relapsing-remitting multiple sclerosis   \n",
       "SRR1839793  relapsing-remitting multiple sclerosis   \n",
       "SRR1839794                                  normal   \n",
       "SRR1839795  relapsing-remitting multiple sclerosis   \n",
       "SRR1839796  relapsing-remitting multiple sclerosis   \n",
       "SRR1839797  relapsing-remitting multiple sclerosis   \n",
       "SRR1839798  relapsing-remitting multiple sclerosis   \n",
       "SRR1839799                                  normal   \n",
       "SRR1839800                                  normal   \n",
       "SRR1839801                                  normal   \n",
       "SRR1839802                                  normal   \n",
       "SRR1839803                                  normal   \n",
       "SRR1839804                                  normal   \n",
       "\n",
       "           Sample Characteristic Ontology Term[disease]  \\\n",
       "Run                                                       \n",
       "SRR1839791  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839792         http://www.ebi.ac.uk/efo/EFO_0003929   \n",
       "SRR1839793         http://www.ebi.ac.uk/efo/EFO_0003929   \n",
       "SRR1839794  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839795         http://www.ebi.ac.uk/efo/EFO_0003929   \n",
       "SRR1839796         http://www.ebi.ac.uk/efo/EFO_0003929   \n",
       "SRR1839797         http://www.ebi.ac.uk/efo/EFO_0003929   \n",
       "SRR1839798         http://www.ebi.ac.uk/efo/EFO_0003929   \n",
       "SRR1839799  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839800  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839801  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839802  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839803  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "SRR1839804  http://purl.obolibrary.org/obo/PATO_0000461   \n",
       "\n",
       "           Sample Characteristic[ethnic group]  \\\n",
       "Run                                              \n",
       "SRR1839791                           Caucasian   \n",
       "SRR1839792                           Caucasian   \n",
       "SRR1839793                           Caucasian   \n",
       "SRR1839794                           Caucasian   \n",
       "SRR1839795                           Caucasian   \n",
       "SRR1839796                           Caucasian   \n",
       "SRR1839797                           Caucasian   \n",
       "SRR1839798                           Caucasian   \n",
       "SRR1839799                           Caucasian   \n",
       "SRR1839800                           Caucasian   \n",
       "SRR1839801                           Caucasian   \n",
       "SRR1839802                           Caucasian   \n",
       "SRR1839803                           Caucasian   \n",
       "SRR1839804                           Caucasian   \n",
       "\n",
       "           Sample Characteristic Ontology Term[ethnic group]  \\\n",
       "Run                                                            \n",
       "SRR1839791     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839792     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839793     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839794     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839795     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839796     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839797     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839798     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839799     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839800     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839801     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839802     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839803     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "SRR1839804     http://purl.obolibrary.org/obo/HANCESTRO_0005   \n",
       "\n",
       "           Sample Characteristic[organism]  \\\n",
       "Run                                          \n",
       "SRR1839791                    Homo sapiens   \n",
       "SRR1839792                    Homo sapiens   \n",
       "SRR1839793                    Homo sapiens   \n",
       "SRR1839794                    Homo sapiens   \n",
       "SRR1839795                    Homo sapiens   \n",
       "SRR1839796                    Homo sapiens   \n",
       "SRR1839797                    Homo sapiens   \n",
       "SRR1839798                    Homo sapiens   \n",
       "SRR1839799                    Homo sapiens   \n",
       "SRR1839800                    Homo sapiens   \n",
       "SRR1839801                    Homo sapiens   \n",
       "SRR1839802                    Homo sapiens   \n",
       "SRR1839803                    Homo sapiens   \n",
       "SRR1839804                    Homo sapiens   \n",
       "\n",
       "            Sample Characteristic Ontology Term[organism]  \\\n",
       "Run                                                         \n",
       "SRR1839791  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839792  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839793  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839794  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839795  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839796  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839797  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839798  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839799  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839800  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839801  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839802  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839803  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "SRR1839804  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "\n",
       "           Sample Characteristic[organism part]  \\\n",
       "Run                                               \n",
       "SRR1839791                                blood   \n",
       "SRR1839792                                blood   \n",
       "SRR1839793                                blood   \n",
       "SRR1839794                                blood   \n",
       "SRR1839795                                blood   \n",
       "SRR1839796                                blood   \n",
       "SRR1839797                                blood   \n",
       "SRR1839798                                blood   \n",
       "SRR1839799                                blood   \n",
       "SRR1839800                                blood   \n",
       "SRR1839801                                blood   \n",
       "SRR1839802                                blood   \n",
       "SRR1839803                                blood   \n",
       "SRR1839804                                blood   \n",
       "\n",
       "           Sample Characteristic Ontology Term[organism part]  \\\n",
       "Run                                                             \n",
       "SRR1839791      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839792      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839793      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839794      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839795      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839796      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839797      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839798      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839799      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839800      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839801      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839802      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839803      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "SRR1839804      http://purl.obolibrary.org/obo/UBERON_0000178   \n",
       "\n",
       "                             Factor Value[disease]  \\\n",
       "Run                                                  \n",
       "SRR1839791                                  normal   \n",
       "SRR1839792  relapsing-remitting multiple sclerosis   \n",
       "SRR1839793  relapsing-remitting multiple sclerosis   \n",
       "SRR1839794                                  normal   \n",
       "SRR1839795  relapsing-remitting multiple sclerosis   \n",
       "SRR1839796  relapsing-remitting multiple sclerosis   \n",
       "SRR1839797  relapsing-remitting multiple sclerosis   \n",
       "SRR1839798  relapsing-remitting multiple sclerosis   \n",
       "SRR1839799                                  normal   \n",
       "SRR1839800                                  normal   \n",
       "SRR1839801                                  normal   \n",
       "SRR1839802                                  normal   \n",
       "SRR1839803                                  normal   \n",
       "SRR1839804                                  normal   \n",
       "\n",
       "                    Factor Value Ontology Term[disease] Analysed  \n",
       "Run                                                               \n",
       "SRR1839791  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839792         http://www.ebi.ac.uk/efo/EFO_0003929      Yes  \n",
       "SRR1839793         http://www.ebi.ac.uk/efo/EFO_0003929      Yes  \n",
       "SRR1839794  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839795         http://www.ebi.ac.uk/efo/EFO_0003929      Yes  \n",
       "SRR1839796         http://www.ebi.ac.uk/efo/EFO_0003929      Yes  \n",
       "SRR1839797         http://www.ebi.ac.uk/efo/EFO_0003929      Yes  \n",
       "SRR1839798         http://www.ebi.ac.uk/efo/EFO_0003929      Yes  \n",
       "SRR1839799  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839800  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839801  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839802  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839803  http://purl.obolibrary.org/obo/PATO_0000461      Yes  \n",
       "SRR1839804  http://purl.obolibrary.org/obo/PATO_0000461      Yes  "
      ]
     },
     "execution_count": 283,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_experiment_design = pd.read_csv(f\"{DATA_DIR}/experiment-design.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_experiment_design\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6g7h8i9",
   "metadata": {},
   "source": [
    "### Query Results\n",
    "\n",
    "This file contains the query results from the Expression Atlas database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "ac40d284",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>'relapsing-remitting multiple sclerosis' vs 'normal' .foldChange</th>\n",
       "      <th>'relapsing-remitting multiple sclerosis' vs 'normal'.pValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.686278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.750671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>0.348980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>FGR</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.099978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>CFH</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>0.536052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285966</th>\n",
       "      <td>ENSG00000285966</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.070963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285967</th>\n",
       "      <td>NIPBL-DT</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.005982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285976</th>\n",
       "      <td>ENSG00000285976</td>\n",
       "      <td>-1.1</td>\n",
       "      <td>0.042264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285979</th>\n",
       "      <td>ENSG00000285979</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>0.134743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285994</th>\n",
       "      <td>ENSG00000285994</td>\n",
       "      <td>-1.3</td>\n",
       "      <td>0.017190</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14310 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene Name  \\\n",
       "Gene ID                            \n",
       "ENSG00000000419             DPM1   \n",
       "ENSG00000000457            SCYL3   \n",
       "ENSG00000000460         C1orf112   \n",
       "ENSG00000000938              FGR   \n",
       "ENSG00000000971              CFH   \n",
       "...                          ...   \n",
       "ENSG00000285966  ENSG00000285966   \n",
       "ENSG00000285967         NIPBL-DT   \n",
       "ENSG00000285976  ENSG00000285976   \n",
       "ENSG00000285979  ENSG00000285979   \n",
       "ENSG00000285994  ENSG00000285994   \n",
       "\n",
       "                 'relapsing-remitting multiple sclerosis' vs 'normal' .foldChange  \\\n",
       "Gene ID                                                                             \n",
       "ENSG00000000419                                                0.1                  \n",
       "ENSG00000000457                                                0.1                  \n",
       "ENSG00000000460                                               -0.4                  \n",
       "ENSG00000000938                                                0.4                  \n",
       "ENSG00000000971                                               -0.4                  \n",
       "...                                                            ...                  \n",
       "ENSG00000285966                                                0.9                  \n",
       "ENSG00000285967                                               -1.0                  \n",
       "ENSG00000285976                                               -1.1                  \n",
       "ENSG00000285979                                               -0.6                  \n",
       "ENSG00000285994                                               -1.3                  \n",
       "\n",
       "                 'relapsing-remitting multiple sclerosis' vs 'normal'.pValue  \n",
       "Gene ID                                                                       \n",
       "ENSG00000000419                                           0.686278            \n",
       "ENSG00000000457                                           0.750671            \n",
       "ENSG00000000460                                           0.348980            \n",
       "ENSG00000000938                                           0.099978            \n",
       "ENSG00000000971                                           0.536052            \n",
       "...                                                            ...            \n",
       "ENSG00000285966                                           0.070963            \n",
       "ENSG00000285967                                           0.005982            \n",
       "ENSG00000285976                                           0.042264            \n",
       "ENSG00000285979                                           0.134743            \n",
       "ENSG00000285994                                           0.017190            \n",
       "\n",
       "[14310 rows x 3 columns]"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_query_results = pd.read_csv(f\"{DATA_DIR}/query-results.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_query_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "g7h8i9j0",
   "metadata": {},
   "source": [
    "### Data Preparation\n",
    "\n",
    "Before we can train a classifier, we need to prepare the data. This involves selecting features and labels from the dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "49036aca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[8,\n",
       " 135,\n",
       " 303,\n",
       " 46,\n",
       " 14749,\n",
       " 74,\n",
       " 251,\n",
       " 336,\n",
       " 707,\n",
       " 58,\n",
       " 774,\n",
       " 433,\n",
       " 242,\n",
       " 1,\n",
       " 0,\n",
       " 585,\n",
       " 9,\n",
       " 165,\n",
       " 179,\n",
       " 0,\n",
       " 137,\n",
       " 526,\n",
       " 1604,\n",
       " 18,\n",
       " 88,\n",
       " 14,\n",
       " 0,\n",
       " 599,\n",
       " 11091,\n",
       " 1427,\n",
       " 1304,\n",
       " 1517,\n",
       " 14,\n",
       " 63,\n",
       " 60,\n",
       " 90,\n",
       " 189,\n",
       " 1084,\n",
       " 12565,\n",
       " 9,\n",
       " 182,\n",
       " 4015,\n",
       " 28,\n",
       " 2,\n",
       " 1840,\n",
       " 249,\n",
       " 724,\n",
       " 1817,\n",
       " 972,\n",
       " 355,\n",
       " 534,\n",
       " 770,\n",
       " 1975,\n",
       " 682,\n",
       " 298,\n",
       " 185,\n",
       " 0,\n",
       " 180,\n",
       " 162,\n",
       " 321,\n",
       " 8,\n",
       " 54,\n",
       " 0,\n",
       " 143,\n",
       " 65,\n",
       " 523,\n",
       " 2875,\n",
       " 208,\n",
       " 562,\n",
       " 9,\n",
       " 5530,\n",
       " 2485,\n",
       " 1537,\n",
       " 508,\n",
       " 0,\n",
       " 186,\n",
       " 271,\n",
       " 45,\n",
       " 5,\n",
       " 373,\n",
       " 288,\n",
       " 64,\n",
       " 6,\n",
       " 323,\n",
       " 993,\n",
       " 2307,\n",
       " 17,\n",
       " 590,\n",
       " 1822,\n",
       " 6909,\n",
       " 880,\n",
       " 251,\n",
       " 131,\n",
       " 178,\n",
       " 155,\n",
       " 20,\n",
       " 2860,\n",
       " 901,\n",
       " 81,\n",
       " 380,\n",
       " 154,\n",
       " 2052,\n",
       " 1073,\n",
       " 10761,\n",
       " 237,\n",
       " 103,\n",
       " 785,\n",
       " 6022,\n",
       " 647,\n",
       " 1140,\n",
       " 801,\n",
       " 14,\n",
       " 553,\n",
       " 1,\n",
       " 1,\n",
       " 1323,\n",
       " 0,\n",
       " 8,\n",
       " 3054,\n",
       " 575,\n",
       " 472,\n",
       " 74,\n",
       " 115,\n",
       " 332,\n",
       " 27,\n",
       " 2131,\n",
       " 1,\n",
       " 541,\n",
       " 1292,\n",
       " 292,\n",
       " 436,\n",
       " 0,\n",
       " 227,\n",
       " 43,\n",
       " 118,\n",
       " 51,\n",
       " 1572,\n",
       " 0,\n",
       " 195,\n",
       " 3,\n",
       " 937,\n",
       " 915,\n",
       " 15,\n",
       " 798,\n",
       " 9,\n",
       " 407,\n",
       " 41,\n",
       " 1152,\n",
       " 7,\n",
       " 2,\n",
       " 37,\n",
       " 581,\n",
       " 6,\n",
       " 406,\n",
       " 349,\n",
       " 2448,\n",
       " 2,\n",
       " 2,\n",
       " 2242,\n",
       " 0,\n",
       " 2565,\n",
       " 232,\n",
       " 890,\n",
       " 2,\n",
       " 716,\n",
       " 4,\n",
       " 63,\n",
       " 125,\n",
       " 0,\n",
       " 408,\n",
       " 1,\n",
       " 812,\n",
       " 258,\n",
       " 816,\n",
       " 549,\n",
       " 485,\n",
       " 345,\n",
       " 3,\n",
       " 0,\n",
       " 336,\n",
       " 1819,\n",
       " 5,\n",
       " 204,\n",
       " 686,\n",
       " 9,\n",
       " 2100,\n",
       " 337,\n",
       " 0,\n",
       " 125,\n",
       " 11624,\n",
       " 115,\n",
       " 345,\n",
       " 95,\n",
       " 674,\n",
       " 440,\n",
       " 2274,\n",
       " 97,\n",
       " 23,\n",
       " 36,\n",
       " 68,\n",
       " 187,\n",
       " 94,\n",
       " 231,\n",
       " 746,\n",
       " 330,\n",
       " 3066,\n",
       " 32577,\n",
       " 2812,\n",
       " 1063,\n",
       " 14,\n",
       " 893,\n",
       " 230,\n",
       " 853,\n",
       " 964,\n",
       " 602,\n",
       " 6035,\n",
       " 1060,\n",
       " 541,\n",
       " 170,\n",
       " 72,\n",
       " 204,\n",
       " 3010,\n",
       " 107,\n",
       " 289,\n",
       " 0,\n",
       " 1199,\n",
       " 536,\n",
       " 93,\n",
       " 166,\n",
       " 423,\n",
       " 66,\n",
       " 2842,\n",
       " 1211,\n",
       " 243,\n",
       " 173,\n",
       " 3,\n",
       " 1291,\n",
       " 1526,\n",
       " 96,\n",
       " 71,\n",
       " 16,\n",
       " 2699,\n",
       " 2001,\n",
       " 266,\n",
       " 8,\n",
       " 6412,\n",
       " 2,\n",
       " 582,\n",
       " 5524,\n",
       " 73,\n",
       " 1016,\n",
       " 166,\n",
       " 210,\n",
       " 3337,\n",
       " 1285,\n",
       " 0,\n",
       " 531,\n",
       " 1086,\n",
       " 535,\n",
       " 197,\n",
       " 0,\n",
       " 112,\n",
       " 579,\n",
       " 632,\n",
       " 125,\n",
       " 1876,\n",
       " 2,\n",
       " 1080,\n",
       " 596,\n",
       " 294,\n",
       " 1831,\n",
       " 363,\n",
       " 4542,\n",
       " 1,\n",
       " 0,\n",
       " 236,\n",
       " 532,\n",
       " 2172,\n",
       " 6,\n",
       " 860,\n",
       " 595,\n",
       " 1,\n",
       " 113,\n",
       " 497,\n",
       " 466,\n",
       " 70,\n",
       " 64,\n",
       " 8857,\n",
       " 183,\n",
       " 0,\n",
       " 151,\n",
       " 506,\n",
       " 1054,\n",
       " 42,\n",
       " 97,\n",
       " 243,\n",
       " 645,\n",
       " 2795,\n",
       " 4505,\n",
       " 9000,\n",
       " 0,\n",
       " 3209,\n",
       " 390,\n",
       " 264,\n",
       " 288,\n",
       " 808,\n",
       " 1877,\n",
       " 0,\n",
       " 0,\n",
       " 11220,\n",
       " 7210,\n",
       " 1486,\n",
       " 142,\n",
       " 39,\n",
       " 3060,\n",
       " 103,\n",
       " 539,\n",
       " 308,\n",
       " 352,\n",
       " 469,\n",
       " 234,\n",
       " 0,\n",
       " 13,\n",
       " 4192,\n",
       " 3238,\n",
       " 410,\n",
       " 102,\n",
       " 949,\n",
       " 3140,\n",
       " 425,\n",
       " 268,\n",
       " 319,\n",
       " 283,\n",
       " 421,\n",
       " 3228,\n",
       " 212,\n",
       " 606,\n",
       " 9601,\n",
       " 2,\n",
       " 1646,\n",
       " 70,\n",
       " 2,\n",
       " 505,\n",
       " 1,\n",
       " 0,\n",
       " 1163,\n",
       " 14,\n",
       " 0,\n",
       " 0,\n",
       " 408,\n",
       " 604,\n",
       " 0,\n",
       " 503,\n",
       " 2412,\n",
       " 297,\n",
       " 16733,\n",
       " 8,\n",
       " 441,\n",
       " 147,\n",
       " 100,\n",
       " 25,\n",
       " 112,\n",
       " 12,\n",
       " 61,\n",
       " 0,\n",
       " 177,\n",
       " 38467,\n",
       " 113,\n",
       " 312,\n",
       " 476,\n",
       " 119,\n",
       " 176,\n",
       " 32,\n",
       " 70,\n",
       " 4113,\n",
       " 544,\n",
       " 429,\n",
       " 2729,\n",
       " 0,\n",
       " 0,\n",
       " 629,\n",
       " 0,\n",
       " 1221,\n",
       " 869,\n",
       " 5,\n",
       " 0,\n",
       " 264,\n",
       " 2121,\n",
       " 111,\n",
       " 4161,\n",
       " 5508,\n",
       " 287,\n",
       " 552,\n",
       " 312,\n",
       " 2484,\n",
       " 808,\n",
       " 903,\n",
       " 979,\n",
       " 715,\n",
       " 1396,\n",
       " 923,\n",
       " 235,\n",
       " 31,\n",
       " 168,\n",
       " 616,\n",
       " 70,\n",
       " 4491,\n",
       " 4156,\n",
       " 236,\n",
       " 3858,\n",
       " 16,\n",
       " 2,\n",
       " 506,\n",
       " 342,\n",
       " 121,\n",
       " 1409,\n",
       " 10,\n",
       " 105,\n",
       " 7086,\n",
       " 546,\n",
       " 260,\n",
       " 634,\n",
       " 1332,\n",
       " 8620,\n",
       " 70,\n",
       " 496,\n",
       " 8772,\n",
       " 7605,\n",
       " 22,\n",
       " 252,\n",
       " 1046,\n",
       " 4619,\n",
       " 51,\n",
       " 2245,\n",
       " 3,\n",
       " 2135,\n",
       " 304,\n",
       " 658,\n",
       " 81,\n",
       " 22425,\n",
       " 226,\n",
       " 3220,\n",
       " 672,\n",
       " 2405,\n",
       " 190,\n",
       " 10,\n",
       " 1751,\n",
       " 990,\n",
       " 385,\n",
       " 67,\n",
       " 719,\n",
       " 7,\n",
       " 689,\n",
       " 543,\n",
       " 0,\n",
       " 398,\n",
       " 13696,\n",
       " 742,\n",
       " 260,\n",
       " 14,\n",
       " 1457,\n",
       " 1473,\n",
       " 682,\n",
       " 209,\n",
       " 2554,\n",
       " 67,\n",
       " 168,\n",
       " 527,\n",
       " 189,\n",
       " 1981,\n",
       " 19,\n",
       " 162,\n",
       " 538,\n",
       " 4714,\n",
       " 1684,\n",
       " 1749,\n",
       " 530,\n",
       " 687,\n",
       " 4315,\n",
       " 16,\n",
       " 2136,\n",
       " 202,\n",
       " 789,\n",
       " 65,\n",
       " 1285,\n",
       " 381,\n",
       " 257,\n",
       " 3237,\n",
       " 3,\n",
       " 2071,\n",
       " 773,\n",
       " 860,\n",
       " 52,\n",
       " 5026,\n",
       " 590,\n",
       " 527,\n",
       " 1290,\n",
       " 49,\n",
       " 1,\n",
       " 11,\n",
       " 370,\n",
       " 11,\n",
       " 3,\n",
       " 0,\n",
       " 101,\n",
       " 4,\n",
       " 222,\n",
       " 938,\n",
       " 726,\n",
       " 278,\n",
       " 384,\n",
       " 45,\n",
       " 0,\n",
       " 122,\n",
       " 137,\n",
       " 3971,\n",
       " 2341,\n",
       " 1130,\n",
       " 492,\n",
       " 9870,\n",
       " 806,\n",
       " 25,\n",
       " 3,\n",
       " 432,\n",
       " 0,\n",
       " 953,\n",
       " 1689,\n",
       " 4,\n",
       " 6,\n",
       " 973,\n",
       " 5,\n",
       " 158,\n",
       " 45,\n",
       " 260,\n",
       " 188,\n",
       " 285,\n",
       " 50,\n",
       " 1074,\n",
       " 0,\n",
       " 2950,\n",
       " 160,\n",
       " 285,\n",
       " 5,\n",
       " 491,\n",
       " 168,\n",
       " 5,\n",
       " 159,\n",
       " 55,\n",
       " 321,\n",
       " 135,\n",
       " 0,\n",
       " 11,\n",
       " 177,\n",
       " 461,\n",
       " 1065,\n",
       " 671,\n",
       " 0,\n",
       " 0,\n",
       " 37,\n",
       " 406,\n",
       " 0,\n",
       " 362,\n",
       " 2285,\n",
       " 5737,\n",
       " 202,\n",
       " 2,\n",
       " 331,\n",
       " 1805,\n",
       " 1086,\n",
       " 36,\n",
       " 1,\n",
       " 3103,\n",
       " 3,\n",
       " 139,\n",
       " 765,\n",
       " 60,\n",
       " 0,\n",
       " 1261,\n",
       " 559,\n",
       " 81,\n",
       " 393,\n",
       " 1204,\n",
       " 865,\n",
       " 811,\n",
       " 2431,\n",
       " 5,\n",
       " 830,\n",
       " 217,\n",
       " 106,\n",
       " 2,\n",
       " 127,\n",
       " 111,\n",
       " 2732,\n",
       " 45,\n",
       " 31,\n",
       " 1224,\n",
       " 444,\n",
       " 663,\n",
       " 209,\n",
       " 547,\n",
       " 96,\n",
       " 69,\n",
       " 1031,\n",
       " 601,\n",
       " 5,\n",
       " 970,\n",
       " 0,\n",
       " 224,\n",
       " 1,\n",
       " 675,\n",
       " 1419,\n",
       " 8272,\n",
       " 3829,\n",
       " 420,\n",
       " 276,\n",
       " 2,\n",
       " 56,\n",
       " 27,\n",
       " 1628,\n",
       " 1413,\n",
       " 87,\n",
       " 0,\n",
       " 189,\n",
       " 0,\n",
       " 73,\n",
       " 52,\n",
       " 0,\n",
       " 263,\n",
       " 1368,\n",
       " 1251,\n",
       " 109,\n",
       " 159,\n",
       " 121,\n",
       " 1084,\n",
       " 149,\n",
       " 14,\n",
       " 400,\n",
       " 42,\n",
       " 19,\n",
       " 293,\n",
       " 205,\n",
       " 147,\n",
       " 933,\n",
       " 126,\n",
       " 10,\n",
       " 5,\n",
       " 11,\n",
       " 404,\n",
       " 4,\n",
       " 2,\n",
       " 1472,\n",
       " 1758,\n",
       " 276,\n",
       " 2,\n",
       " 17,\n",
       " 744,\n",
       " 10983,\n",
       " 23,\n",
       " 1036,\n",
       " 359,\n",
       " 33,\n",
       " 1,\n",
       " 549,\n",
       " 1370,\n",
       " 189,\n",
       " 134,\n",
       " 1105,\n",
       " 0,\n",
       " 3,\n",
       " 1999,\n",
       " 4,\n",
       " 357,\n",
       " 206,\n",
       " 40,\n",
       " 215,\n",
       " 24,\n",
       " 89,\n",
       " 1,\n",
       " 547,\n",
       " 404,\n",
       " 1166,\n",
       " 523,\n",
       " 3150,\n",
       " 251,\n",
       " 20,\n",
       " 568,\n",
       " 1206,\n",
       " 62,\n",
       " 131,\n",
       " 13,\n",
       " 68,\n",
       " 1390,\n",
       " 3,\n",
       " 1078,\n",
       " 3646,\n",
       " 2,\n",
       " 135,\n",
       " 0,\n",
       " 1885,\n",
       " 1626,\n",
       " 714,\n",
       " 233,\n",
       " 1938,\n",
       " 84,\n",
       " 622,\n",
       " 162,\n",
       " 5530,\n",
       " 2406,\n",
       " 322,\n",
       " 1163,\n",
       " 1109,\n",
       " 2714,\n",
       " 12,\n",
       " 0,\n",
       " 3248,\n",
       " 263,\n",
       " 22,\n",
       " 12,\n",
       " 58,\n",
       " 1034,\n",
       " 71,\n",
       " 0,\n",
       " 1836,\n",
       " 1161,\n",
       " 53,\n",
       " 211,\n",
       " 3,\n",
       " 32,\n",
       " 1486,\n",
       " 16,\n",
       " 1,\n",
       " 2,\n",
       " 2408,\n",
       " 911,\n",
       " 163,\n",
       " 2100,\n",
       " 524,\n",
       " 170,\n",
       " 115,\n",
       " 1826,\n",
       " 280,\n",
       " 3121,\n",
       " 1734,\n",
       " 1,\n",
       " 1,\n",
       " 765,\n",
       " 573,\n",
       " 2454,\n",
       " 12,\n",
       " 127,\n",
       " 514,\n",
       " 271,\n",
       " 209,\n",
       " 1155,\n",
       " 1100,\n",
       " 3536,\n",
       " 1153,\n",
       " 344,\n",
       " 492,\n",
       " 102,\n",
       " 18621,\n",
       " 627,\n",
       " 48,\n",
       " 6511,\n",
       " 61,\n",
       " 1260,\n",
       " 4189,\n",
       " 28,\n",
       " 5214,\n",
       " 621,\n",
       " 3053,\n",
       " 438,\n",
       " 0,\n",
       " 258,\n",
       " 15,\n",
       " 326,\n",
       " 0,\n",
       " 0,\n",
       " 375,\n",
       " 217,\n",
       " 1543,\n",
       " 528,\n",
       " 3620,\n",
       " 7,\n",
       " 0,\n",
       " 11,\n",
       " 55,\n",
       " 213,\n",
       " 95,\n",
       " 1146,\n",
       " 4618,\n",
       " 1060,\n",
       " 0,\n",
       " 1151,\n",
       " 3703,\n",
       " 2,\n",
       " 1500,\n",
       " 459,\n",
       " 10,\n",
       " 1457,\n",
       " 1469,\n",
       " 6931,\n",
       " 539,\n",
       " 428,\n",
       " 13,\n",
       " 3249,\n",
       " 73,\n",
       " 923,\n",
       " 300,\n",
       " 7075,\n",
       " 116,\n",
       " 107,\n",
       " 3654,\n",
       " 2333,\n",
       " 911,\n",
       " 7,\n",
       " 416,\n",
       " 1062,\n",
       " 15,\n",
       " 161,\n",
       " 1041,\n",
       " 1811,\n",
       " 3448,\n",
       " 1,\n",
       " 150,\n",
       " 488,\n",
       " 0,\n",
       " 3,\n",
       " 1722,\n",
       " 0,\n",
       " 338,\n",
       " 13,\n",
       " 9,\n",
       " 8,\n",
       " 441,\n",
       " 1223,\n",
       " 1213,\n",
       " 7,\n",
       " 430,\n",
       " 200,\n",
       " 3922,\n",
       " 3032,\n",
       " 1919,\n",
       " 136,\n",
       " 31,\n",
       " 0,\n",
       " 12195,\n",
       " 7662,\n",
       " 0,\n",
       " 277,\n",
       " 653,\n",
       " 284,\n",
       " 45,\n",
       " 4,\n",
       " 133,\n",
       " 9157,\n",
       " 162,\n",
       " 770,\n",
       " 38,\n",
       " 291,\n",
       " 831,\n",
       " 2666,\n",
       " 759,\n",
       " 54,\n",
       " 60,\n",
       " 341,\n",
       " 1492,\n",
       " 871,\n",
       " 398,\n",
       " 260,\n",
       " 1920,\n",
       " 324,\n",
       " 543,\n",
       " 8,\n",
       " 4,\n",
       " 3716,\n",
       " 6,\n",
       " 4356,\n",
       " 1121,\n",
       " 364,\n",
       " 72,\n",
       " 251,\n",
       " 358,\n",
       " 3398,\n",
       " 106,\n",
       " 220,\n",
       " 981,\n",
       " 89,\n",
       " 1080,\n",
       " 1556,\n",
       " 10,\n",
       " 36,\n",
       " 248,\n",
       " 622,\n",
       " 179,\n",
       " 803,\n",
       " 2340,\n",
       " 26,\n",
       " 2547,\n",
       " 1549,\n",
       " 183,\n",
       " 215,\n",
       " 1103,\n",
       " 4117,\n",
       " 305,\n",
       " 2332,\n",
       " 0,\n",
       " 900,\n",
       " 17,\n",
       " 2067,\n",
       " 1026,\n",
       " 376,\n",
       " 1631,\n",
       " 1,\n",
       " 28,\n",
       " 0,\n",
       " 4,\n",
       " 1308,\n",
       " 1424,\n",
       " 17651,\n",
       " 177,\n",
       " 4,\n",
       " 0,\n",
       " 761,\n",
       " 528,\n",
       " 393,\n",
       " 3,\n",
       " 121,\n",
       " 198,\n",
       " 2,\n",
       " 58,\n",
       " 735,\n",
       " 50,\n",
       " 10,\n",
       " 900,\n",
       " 1584,\n",
       " 0,\n",
       " 270,\n",
       " 187,\n",
       " 110,\n",
       " 338,\n",
       " 429,\n",
       " 0,\n",
       " 723,\n",
       " 379,\n",
       " 1552,\n",
       " 7555,\n",
       " 38,\n",
       " 191,\n",
       " 1976,\n",
       " 2,\n",
       " 9306,\n",
       " 205,\n",
       " 292,\n",
       " 118,\n",
       " 8215,\n",
       " 241,\n",
       " 196,\n",
       " 246,\n",
       " 381,\n",
       " 17,\n",
       " 119,\n",
       " 11622,\n",
       " 1516,\n",
       " 11,\n",
       " 254,\n",
       " 0,\n",
       " 841,\n",
       " 1,\n",
       " 31,\n",
       " 775,\n",
       " 262,\n",
       " 924,\n",
       " 0,\n",
       " 1593,\n",
       " 889,\n",
       " 375,\n",
       " 134,\n",
       " 1402,\n",
       " 2628,\n",
       " 0,\n",
       " 563,\n",
       " 220,\n",
       " 1160,\n",
       " 18,\n",
       " 696,\n",
       " 2794,\n",
       " 3721,\n",
       " 647,\n",
       " 1591,\n",
       " 557,\n",
       " 326,\n",
       " 1497,\n",
       " 428,\n",
       " 79,\n",
       " 1,\n",
       " 2807,\n",
       " 430,\n",
       " 1067,\n",
       " 1636,\n",
       " ...]"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_features_of_patient(df, patient_id):\n",
    "    \"\"\"\n",
    "    Get the raw gene counts of a specific patient from the DataFrame.\n",
    "    \n",
    "    Parameters:\n",
    "    df (pd.DataFrame): DataFrame containing patient data.\n",
    "    patient_id (str): ID of the patient to retrieve features for. You can find the IDs in the first column of the analytics dataframe.\n",
    "    \n",
    "    Returns:\n",
    "    pd.Series: Features of the specified patient.\n",
    "    \"\"\"\n",
    "    return df[patient_id].to_list() \n",
    "\n",
    "get_features_of_patient(df_filtered, \"SRR1839791\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "251a1ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dataset(df_filtered, df_experiment_design): \n",
    "    \"\"\"\n",
    "    Get the entire dataset of features and labels.\n",
    "    \n",
    "    Parameters:\n",
    "    df_raw_counts (pd.DataFrame): DataFrame containing raw gene counts.\n",
    "    df_experiment_design (pd.DataFrame): DataFrame containing experiment design data.\n",
    "    \n",
    "    Returns:\n",
    "    tuple: Features and labels for the dataset.\n",
    "    \"\"\"\n",
    "    patient_ids = df_experiment_design.index.to_list()\n",
    "    features = [get_features_of_patient(df_filtered, patient_id) for patient_id in patient_ids]\n",
    "    labels = df_experiment_design[\"Sample Characteristic[disease]\"].to_list()\n",
    "    unique_labels = list(set(labels))\n",
    "    labels = [unique_labels.index(label) for label in labels] # Convert labels to numbers\n",
    "    return features, labels\n",
    "\n",
    "features, labels = get_dataset(df_filtered, df_experiment_design)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "h8i9j0k1",
   "metadata": {},
   "source": [
    "### Train/Test Split\n",
    "\n",
    "We split the data into a training set and a testing set. The training set is used to train the classifier, and the testing set is used to evaluate its performance. The main bottleneck we have with that project is the lack of enough data. In fact, we have 14 cases in total with 8 and 6 cases respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "864af4ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_features, test_features, train_labels, test_labels = train_test_split(features, labels, test_size=0.5, random_state=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "i9j0k1l2",
   "metadata": {},
   "source": [
    "### Train the Classifier\n",
    "\n",
    "We use a simple Decision Tree Classifier to train the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "eb1759f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "classifier = DecisionTreeClassifier(random_state=2)\n",
    "classifier = classifier.fit(train_features, train_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "j0k1l2m3",
   "metadata": {},
   "source": [
    "### Evaluate the Classifier\n",
    "\n",
    "We use the accuracy score to evaluate the performance of the classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "a18a6cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "predicted_test_labels = classifier.predict(test_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "c7498d9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Baseline accuracy: 0.8571428571428571\n"
     ]
    }
   ],
   "source": [
    "accuracy = accuracy_score(test_labels, predicted_test_labels)\n",
    "\n",
    "print(f\"Baseline accuracy: {accuracy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "01edc038",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test accuracy without PCA: 0.8571\n",
      "Test accuracy with PCA:    0.8571\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Theo\\AppData\\Local\\Schrodinger\\PyMOL2\\lib\\site-packages\\sklearn\\feature_selection\\_univariate_selection.py:110: UserWarning: Features [   26    43    74 ... 33801 33805 33806] are constant.\n",
      "  warnings.warn(\"Features %s are constant.\" % constant_features_idx, UserWarning)\n",
      "c:\\Users\\Theo\\AppData\\Local\\Schrodinger\\PyMOL2\\lib\\site-packages\\sklearn\\feature_selection\\_univariate_selection.py:111: RuntimeWarning: divide by zero encountered in divide\n",
      "  f = msb / msw\n",
      "c:\\Users\\Theo\\AppData\\Local\\Schrodinger\\PyMOL2\\lib\\site-packages\\sklearn\\feature_selection\\_univariate_selection.py:111: RuntimeWarning: invalid value encountered in divide\n",
      "  f = msb / msw\n"
     ]
    }
   ],
   "source": [
    "# Convert to numpy arrays\n",
    "X_train = np.array(train_features)\n",
    "X_test  = np.array(test_features)\n",
    "y_train = np.array(train_labels)\n",
    "y_test  = np.array(test_labels)\n",
    "\n",
    "# ---- Select top-k genes (ANOVA F-test) ----\n",
    "k = 50\n",
    "selector = SelectKBest(score_func=f_classif, k=min(k, X_train.shape[1]))\n",
    "X_train_selected = selector.fit_transform(X_train, y_train)\n",
    "X_test_selected  = selector.transform(X_test)  # apply same selection to test set\n",
    "\n",
    "# Optional: save selected gene indices\n",
    "selected_gene_indices = selector.get_support(indices=True)\n",
    "selected_genes = df_filtered.index[selected_gene_indices].tolist()\n",
    "\n",
    "# ---- PCA on training data only ----\n",
    "scaler = StandardScaler()\n",
    "X_train_scaled = scaler.fit_transform(X_train_selected)\n",
    "X_test_scaled  = scaler.transform(X_test_selected)\n",
    "\n",
    "pca = PCA(n_components=4)\n",
    "X_train_pca = pca.fit_transform(X_train_scaled)\n",
    "X_test_pca  = pca.transform(X_test_scaled)\n",
    "\n",
    "# ---- Model without PCA ----\n",
    "model_no_pca = DecisionTreeClassifier(random_state=42)\n",
    "model_no_pca.fit(X_train_scaled, y_train)\n",
    "y_pred_no_pca = model_no_pca.predict(X_test_scaled)\n",
    "accuracy_no_pca = accuracy_score(y_test, y_pred_no_pca)\n",
    "\n",
    "# ---- Model with PCA ----\n",
    "model_pca = DecisionTreeClassifier(random_state=42)\n",
    "model_pca.fit(X_train_pca, y_train)\n",
    "y_pred_pca = model_pca.predict(X_test_pca)\n",
    "accuracy_pca = accuracy_score(y_test, y_pred_pca)\n",
    "\n",
    "print(f\"Test accuracy without PCA: {accuracy_no_pca:.4f}\")\n",
    "print(f\"Test accuracy with PCA:    {accuracy_pca:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "covariance_matrix_cell",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Explained variance of all PCs: [0.51632513 0.25227881 0.11642284 0.04280668]\n",
      "Mean Accuracy: 0.929\n",
      "Standard Deviation: 0.175\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "# Load the data\n",
    "data = np.array(features)\n",
    "\n",
    "# Transpose the data so that rows are samples and columns are features (genes)\n",
    "data = data\n",
    "\n",
    "kf = KFold(n_splits=7, shuffle=True, random_state=42)\n",
    "accuracies = []\n",
    "\n",
    "patient_ids = df_experiment_design.index.to_numpy()\n",
    "labels = np.array(labels)\n",
    "\n",
    "labels = np.array(labels)\n",
    "\n",
    "for train_idx, test_idx in kf.split(patient_ids):\n",
    "    \n",
    "    train_samples = patient_ids[train_idx]\n",
    "    test_samples = patient_ids[test_idx]\n",
    "    \n",
    "    healthy_train = df_experiment_design.loc[train_samples, \"Sample Characteristic[disease]\"] == \"normal\"\n",
    "    disease_train = ~healthy_train\n",
    "    \n",
    "    mean_h = df_raw_counts[train_samples[healthy_train]].mean(axis=1)\n",
    "    mean_d = df_raw_counts[train_samples[disease_train]].mean(axis=1)\n",
    "    \n",
    "    fold = (mean_d - mean_h).abs()\n",
    "    top_genes = fold.nlargest(50).index\n",
    "    \n",
    "    X_train = df_raw_counts.loc[top_genes, train_samples].T.values\n",
    "    X_test = df_raw_counts.loc[top_genes, test_samples].T.values\n",
    "    \n",
    "    y_train = labels[train_idx]\n",
    "    y_test = labels[test_idx]\n",
    "    \n",
    "    scaler = StandardScaler()\n",
    "    X_train = scaler.fit_transform(X_train)\n",
    "    X_test = scaler.transform(X_test)\n",
    "    \n",
    "    pca = PCA(n_components=4)\n",
    "    X_train_pca = pca.fit_transform(X_train)\n",
    "    X_test_pca = pca.transform(X_test)\n",
    "    \n",
    "    #model = DecisionTreeClassifier(random_state=42)\n",
    "    #model = LogisticRegression(penalty='l2', solver='liblinear', random_state=42)\n",
    "    #model = KNeighborsClassifier(n_neighbors=3)\n",
    "    #model = LogisticRegression(penalty='l1', solver='liblinear', C=1.0, random_state=42)\n",
    "    model = SVC(kernel='linear', C=1.0, random_state=42)\n",
    "    model.fit(X_train_pca, y_train)\n",
    "    \n",
    "    y_pred = model.predict(X_test_pca)\n",
    "    accuracies.append(accuracy_score(y_test, y_pred))\n",
    "\n",
    "print(\"Explained variance of all PCs:\", pca.explained_variance_ratio_)\n",
    "print(f\"Mean Accuracy: {np.mean(accuracies):.3f}\")\n",
    "print(f\"Standard Deviation: {np.std(accuracies):.3f}\")\n",
    "\n",
    "\n",
    "# Plot the covariance matrix\n",
    "#plt.figure(figsize=(10, 8))\n",
    "#plt.imshow(covariance_matrix, cmap='viridis', interpolation='nearest')\n",
    "#plt.title('Covariance Matrix of Features')\n",
    "#plt.colorbar()\n",
    "#plt.xlabel('Features')\n",
    "#plt.ylabel('Features')\n",
    "#plt.savefig('covariance_matrix.png')\n",
    "#print(\"Covariance matrix plot saved to covariance_matrix.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "0971fce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_single_gene_feature(df_filtered, df_experiment_design, gene): \n",
    "    \"\"\"\n",
    "    Get the single gene feature for all patients in the dataset.\n",
    "    \n",
    "    Parameters:\n",
    "    df_filtered (pd.DataFrame): DataFrame containing raw gene counts.\n",
    "    df_experiment_design (pd.DataFrame): DataFrame containing experiment design data.\n",
    "    \n",
    "    Returns:\n",
    "    tuple: Features and labels for the dataset.\n",
    "    \"\"\"\n",
    "    patient_ids = df_experiment_design.index.to_list()\n",
    "\n",
    "    gene_index = df_filtered.index.get_loc(gene)\n",
    "\n",
    "    features = [df_filtered[patient_id][gene_index:gene_index + 1].to_list() for patient_id in patient_ids]\n",
    "    labels = df_experiment_design[\"Sample Characteristic[disease]\"].to_list()\n",
    "    unique_labels = list(set(labels))\n",
    "    labels = [unique_labels.index(label) for label in labels] # Convert labels to numbers\n",
    "    return features, labels\n",
    "\n",
    "single_feature, labels = get_single_gene_feature(df_filtered, df_experiment_design, \"ENSG00000250696\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "7a8a6a89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjkAAAGdCAYAAADwjmIIAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjcsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvTLEjVAAAAAlwSFlzAAAPYQAAD2EBqD+naQAAI5dJREFUeJzt3XtwVOXh//HPJiEBhSRyyxJIuFgUEAiamLBiR78mNV6mNRVbpCiIjPyk4aKhjqBAbK2NyqASQVLotOgAhcYWClTRNGi0JdwSrFwjWgQEN5FidjFIErPn9wdldcvmSpZNHt6vmZ025zxn9zlP2O57Nme3NsuyLAEAABgmJNgTAAAACAQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRwoI9gWDweDw6fvy4unTpIpvNFuzpAACAJrAsS6dOnVJsbKxCQhp/n+aSjJzjx48rLi4u2NMAAAAtcPToUfXp06fRcZdk5HTp0kXS2UWKjIwM8mwAAEBTuN1uxcXFeV/HG3NJRs65P1FFRkYSOQAAtDNNvdSEC48BAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGOmiRM7ixYvVr18/dezYUSkpKdq+fXuD4/Pz8zVo0CB17NhRw4YN0xtvvFHv2Icfflg2m00vvfRSK88aAAC0ZwGPnDVr1igrK0vZ2dkqLS1VQkKC0tPTVVFR4Xf8li1bNHbsWE2aNEm7du1SRkaGMjIytGfPnvPGrl27Vlu3blVsbGygTwMAALQzAY+cF154QQ899JAmTpyoIUOGKC8vT5dddpl+//vf+x2/cOFC3XbbbXrsscc0ePBgPf3007ruuuu0aNEin3HHjh3TtGnTtHLlSnXo0CHQpwEAANqZgEZOTU2NSkpKlJaW9u0DhoQoLS1NxcXFfo8pLi72GS9J6enpPuM9Ho/uv/9+PfbYY7rmmmsanUd1dbXcbrfPDQAAmC2gkXPixAnV1dUpJibGZ3tMTIycTqffY5xOZ6Pjn3vuOYWFhWn69OlNmkdOTo6ioqK8t7i4uGaeCQAAaG/a3aerSkpKtHDhQi1fvlw2m61Jx8yePVsul8t7O3r0aIBnCQAAgi2gkdO9e3eFhoaqvLzcZ3t5ebnsdrvfY+x2e4Pj33//fVVUVCg+Pl5hYWEKCwvT4cOHNXPmTPXr18/vfUZERCgyMtLnBgAAzBbQyAkPD1diYqIKCwu92zwejwoLC+VwOPwe43A4fMZLUkFBgXf8/fffrw8//FAffPCB9xYbG6vHHntMb731VuBOBgAAtCthgX6ArKwsTZgwQUlJSUpOTtZLL72kqqoqTZw4UZI0fvx49e7dWzk5OZKkGTNm6KabbtKCBQt05513avXq1dq5c6eWLl0qSerWrZu6devm8xgdOnSQ3W7X1VdfHejTAQAA7UTAI2fMmDH64osvNG/ePDmdTo0YMUKbNm3yXlx85MgRhYR8+4bSDTfcoFWrVmnOnDl64oknNHDgQK1bt05Dhw4N9FQBAIBBbJZlWcGexMXmdrsVFRUll8vF9TkAALQTzX39bnefrgIAAGgKIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkYgcAABgJCIHAAAYicgBAABGInIAAICRiBwAAGAkIgcAABiJyAEAAEYicgAAgJGIHAAAYCQiBwAAGInIAQAARiJyAACAkS5K5CxevFj9+vVTx44dlZKSou3btzc4Pj8/X4MGDVLHjh01bNgwvfHGG959tbW1evzxxzVs2DBdfvnlio2N1fjx43X8+PFAnwYAAGhHAh45a9asUVZWlrKzs1VaWqqEhASlp6eroqLC7/gtW7Zo7NixmjRpknbt2qWMjAxlZGRoz549kqTTp0+rtLRUc+fOVWlpqf7yl7+orKxMP/rRjwJ9KgAAoB2xWZZlBfIBUlJSdP3112vRokWSJI/Ho7i4OE2bNk2zZs06b/yYMWNUVVWljRs3ereNHDlSI0aMUF5ent/H2LFjh5KTk3X48GHFx8c3Oie3262oqCi5XC5FRka28MwAAMDF1NzX74C+k1NTU6OSkhKlpaV9+4AhIUpLS1NxcbHfY4qLi33GS1J6enq94yXJ5XLJZrMpOjra7/7q6mq53W6fGwAAMFtAI+fEiROqq6tTTEyMz/aYmBg5nU6/xzidzmaNP3PmjB5//HGNHTu23qrLyclRVFSU9xYXF9eCswEAAO1Ju/50VW1trX7605/KsiwtWbKk3nGzZ8+Wy+Xy3o4ePXoRZwkAAIIhLJB33r17d4WGhqq8vNxne3l5uex2u99j7HZ7k8afC5zDhw9r8+bNDf5tLiIiQhERES08CwAA0B4F9J2c8PBwJSYmqrCw0LvN4/GosLBQDofD7zEOh8NnvCQVFBT4jD8XOAcPHtTf//53devWLTAnAAAA2q2AvpMjSVlZWZowYYKSkpKUnJysl156SVVVVZo4caIkafz48erdu7dycnIkSTNmzNBNN92kBQsW6M4779Tq1au1c+dOLV26VNLZwLnnnntUWlqqjRs3qq6uznu9TteuXRUeHh7oUwIAAO1AwCNnzJgx+uKLLzRv3jw5nU6NGDFCmzZt8l5cfOTIEYWEfPuG0g033KBVq1Zpzpw5euKJJzRw4ECtW7dOQ4cOlSQdO3ZM69evlySNGDHC57Heeecd3XzzzYE+JQAA0A4E/Hty2iK+JwcAgPanTX1PDgAAQLAQOQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwUliwJ4BLnKdOOrxF+qpc6hwj9b1BCglt+vHf1Eg7lklffipd0U+6/iEpLDxQs4V04b+zYGrPcw+WxtaMNcV3tbF/DxclchYvXqz58+fL6XQqISFBL7/8spKTk+sdn5+fr7lz5+rTTz/VwIED9dxzz+mOO+7w7rcsS9nZ2Vq2bJkqKys1atQoLVmyRAMHDrwYp3OeFws+UmiITdNTz3/8e5cWS5JWT3acty+38KDqPJYe/cFVAXnsxu5/zG+LFRpi06qHRp6372fLtqrOY2nN/zt/3q02r94HpE2PS+7j3+6MjJVue04a8qPGH+TtuVLxIsnyfGfbHMkxVbr16RbNG43Yt/7CfmfB1J7nHiyNrRlriu9qg/8eAv7nqjVr1igrK0vZ2dkqLS1VQkKC0tPTVVFR4Xf8li1bNHbsWE2aNEm7du1SRkaGMjIytGfPHu+Y559/Xrm5ucrLy9O2bdt0+eWXKz09XWfOnAn06fgVGmLTCwUfKbfwoM/23MKD2vrvk9r675N+973w3xAI1GM3dv+hITZt+eQ/+tmyrT7bf7Zsq7Z88p8Lmltj87rq5DvSn8b7Phkkyf352e371jf8AG/Plbbk+gaOdPbnLbln96N17Vt/Yb+zYGrPcw+Wxtbs7bmsKb7VRp9jNsuyrEA+QEpKiq6//notWrRIkuTxeBQXF6dp06Zp1qxZ540fM2aMqqqqtHHjRu+2kSNHasSIEcrLy5NlWYqNjdXMmTP1i1/8QpLkcrkUExOj5cuX69577210Tm63W1FRUXK5XIqMjGyV8zz34p31g6s0PXWgz8+S6t3n752O1nzsxu7/XNDccGU3rXpo5Hk/B2JeM9Ou1LQPf3z+k8HLdrb+H9nt/23Ob2qkZ2LODxyfuwiVnnTyp6vW4qmTXhra8t9ZMLXnuQdLo2smyRbSwHOQNb2kXMTnWHNfvwP656qamhqVlJRo9uzZ3m0hISFKS0tTcXGx32OKi4uVlZXlsy09PV3r1q2TJB06dEhOp1NpaWne/VFRUUpJSVFxcbHfyKmurlZ1dbX3Z7fbfSGn5de5mHih4CMt2vyxauo850VGQ/sC/dj1+W7Y9Jv1N0lqlcBpaF7TBjilfzTwP56yJPexs3/X7f/983fvWNZw4EiSVXd2nCOz5SeAbx3e0vALXmO/s2Bqz3MPlkbXTI08B1nTS0obfo4F9M9VJ06cUF1dnWJiYny2x8TEyOl0+j3G6XQ2OP7cfzbnPnNychQVFeW9xcXFteh8GjM9daDCQ0NUU+dReGiIT2Q0tC/Qj92Y/w2a1gicBuf1VXnTDq5v3JefNu34po5D4y70dxZM7XnuwdJaa8GaXhra8HPskvgI+ezZs+Vyuby3o0ePBuRxcgsPel/Ma+o8PtejNLQv0I/dGH/X5AR0Xp1jGj9Qqn/cFf2adnxTx6FxF/o7C6b2PPdgaa21YE0vDW34ORbQyOnevbtCQ0NVXu5bb+Xl5bLb7X6PsdvtDY4/95/Nuc+IiAhFRkb63Frbd68/+eiZ25X1g6u8F942tC/Qj92Y716D8+mzd+qGK7v5vRi5Nef18ic9zv59VvVd2GyTInuf/eihP9c/dPZ6gIbYQs+OQ+voe8OF/c6CqT3PPVgaXTP99znImkJt+jkW0MgJDw9XYmKiCgsLvds8Ho8KCwvlcPj/aLLD4fAZL0kFBQXe8f3795fdbvcZ43a7tW3btnrvM9D8Xeg7PXWg90W9oX0XGjqNPXZD9+/vIuNVD41sldBpaF4L/v6J/tb7kf+O/N8nxX9/vu3Z+i9QCws/+zHxhjgyuei4NYWEnv0YqKQW/c6CqT3PPVgaXTPbd56DrOklrw0/xwL+56qsrCwtW7ZMr776qvbv368pU6aoqqpKEydOlCSNHz/e58LkGTNmaNOmTVqwYIEOHDigp556Sjt37tTUqWefUDabTY888oh+/etfa/369dq9e7fGjx+v2NhYZWRkBPp0/KrzWH4v9J2eOlAjB3TVyAFd/e7L+sFVqvNc2IfbGnrsxu6/zmP5vcj4XOhcyNwam9dHXf9P+ulrUmQv3wMjY89ub+w7FW59Wrph+vnv6NhCz27ne3Ja35AfXdjvLJja89yDpbE1u/Vp1hTfaqPPsYB/hFySFi1a5P0ywBEjRig3N1cpKSmSpJtvvln9+vXT8uXLvePz8/M1Z84c75cBPv/8836/DHDp0qWqrKzUjTfeqFdeeUVXXdW0L9ULxEfI0UJ843H708a+0bRZ2vPcg4VvPEZzBPjfQ3Nfvy9K5LQ1RA4AAO1Pc1+/L4lPVwEAgEsPkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASAGLnJMnT2rcuHGKjIxUdHS0Jk2apK+++qrBY86cOaPMzEx169ZNnTt31ujRo1VeXu7d/69//Utjx45VXFycOnXqpMGDB2vhwoWBOgUAANCOBSxyxo0bp71796qgoEAbN27Ue++9p8mTJzd4zKOPPqoNGzYoPz9fRUVFOn78uO6++27v/pKSEvXs2VMrVqzQ3r179eSTT2r27NlatGhRoE4DAAC0UzbLsqzWvtP9+/dryJAh2rFjh5KSkiRJmzZt0h133KHPPvtMsbGx5x3jcrnUo0cPrVq1Svfcc48k6cCBAxo8eLCKi4s1cuRIv4+VmZmp/fv3a/PmzU2en9vtVlRUlFwulyIjI1twhgAA4GJr7ut3QN7JKS4uVnR0tDdwJCktLU0hISHatm2b32NKSkpUW1urtLQ077ZBgwYpPj5excXF9T6Wy+VS165dW2/yAADACGGBuFOn06mePXv6PlBYmLp27Sqn01nvMeHh4YqOjvbZHhMTU+8xW7Zs0Zo1a/S3v/2twflUV1erurra+7Pb7W7CWQAAgPasWe/kzJo1SzabrcHbgQMHAjVXH3v27NFdd92l7Oxs3XrrrQ2OzcnJUVRUlPcWFxd3UeYIAACCp1nv5MycOVMPPPBAg2MGDBggu92uiooKn+3ffPONTp48Kbvd7vc4u92umpoaVVZW+rybU15eft4x+/btU2pqqiZPnqw5c+Y0Ou/Zs2crKyvL+7Pb7SZ0AAAwXLMip0ePHurRo0ej4xwOhyorK1VSUqLExERJ0ubNm+XxeJSSkuL3mMTERHXo0EGFhYUaPXq0JKmsrExHjhyRw+Hwjtu7d69uueUWTZgwQc8880yT5h0REaGIiIgmjQUAAGYIyKerJOn2229XeXm58vLyVFtbq4kTJyopKUmrVq2SJB07dkypqal67bXXlJycLEmaMmWK3njjDS1fvlyRkZGaNm2apLPX3khn/0R1yy23KD09XfPnz/c+VmhoaJPi6xw+XQUAQPvT3NfvgFx4LEkrV67U1KlTlZqaqpCQEI0ePVq5ubne/bW1tSorK9Pp06e921588UXv2OrqaqWnp+uVV17x7n/99df1xRdfaMWKFVqxYoV3e9++ffXpp58G6lQAAEA7FLB3ctoy3skBAKD9aRPfkwMAABBsRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASAGLnJMnT2rcuHGKjIxUdHS0Jk2apK+++qrBY86cOaPMzEx169ZNnTt31ujRo1VeXu537H/+8x/16dNHNptNlZWVATgDAADQngUscsaNG6e9e/eqoKBAGzdu1HvvvafJkyc3eMyjjz6qDRs2KD8/X0VFRTp+/Ljuvvtuv2MnTZqk4cOHB2LqAADAADbLsqzWvtP9+/dryJAh2rFjh5KSkiRJmzZt0h133KHPPvtMsbGx5x3jcrnUo0cPrVq1Svfcc48k6cCBAxo8eLCKi4s1cuRI79glS5ZozZo1mjdvnlJTU/Xll18qOjq6yfNzu92KioqSy+VSZGTkhZ0sAAC4KJr7+h2Qd3KKi4sVHR3tDRxJSktLU0hIiLZt2+b3mJKSEtXW1iotLc27bdCgQYqPj1dxcbF32759+/SrX/1Kr732mkJCmjb96upqud1unxsAADBbQCLH6XSqZ8+ePtvCwsLUtWtXOZ3Oeo8JDw8/7x2ZmJgY7zHV1dUaO3as5s+fr/j4+CbPJycnR1FRUd5bXFxc804IAAC0O82KnFmzZslmszV4O3DgQKDmqtmzZ2vw4MG67777mn2cy+Xy3o4ePRqgGQIAgLYirDmDZ86cqQceeKDBMQMGDJDdbldFRYXP9m+++UYnT56U3W73e5zdbldNTY0qKyt93s0pLy/3HrN582bt3r1br7/+uiTp3OVE3bt315NPPqlf/vKXfu87IiJCERERTTlFAABgiGZFTo8ePdSjR49GxzkcDlVWVqqkpESJiYmSzgaKx+NRSkqK32MSExPVoUMHFRYWavTo0ZKksrIyHTlyRA6HQ5L05z//WV9//bX3mB07dujBBx/U+++/ryuvvLI5pwIAAAzXrMhpqsGDB+u2227TQw89pLy8PNXW1mrq1Km69957vZ+sOnbsmFJTU/Xaa68pOTlZUVFRmjRpkrKystS1a1dFRkZq2rRpcjgc3k9W/W/InDhxwvt4zfl0FQAAMF9AIkeSVq5cqalTpyo1NVUhISEaPXq0cnNzvftra2tVVlam06dPe7e9+OKL3rHV1dVKT0/XK6+8EqgpAgAAgwXke3LaOr4nBwCA9qdNfE8OAABAsBE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIxE5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIYcGeQDBYliVJcrvdQZ4JAABoqnOv2+dexxtzSUbOqVOnJElxcXFBngkAAGiuU6dOKSoqqtFxNqupOWQQj8ej48ePq0uXLrLZbK163263W3FxcTp69KgiIyNb9b5NxZq1DOvWMqxb87FmLcO6tUxD62ZZlk6dOqXY2FiFhDR+xc0l+U5OSEiI+vTpE9DHiIyM5B91M7FmLcO6tQzr1nysWcuwbi1T37o15R2cc7jwGAAAGInIAQAARiJyWllERISys7MVERER7Km0G6xZy7BuLcO6NR9r1jKsW8u05rpdkhceAwAA8/FODgAAMBKRAwAAjETkAAAAIxE5AADASEROK1q8eLH69eunjh07KiUlRdu3bw/2lNqU9957Tz/84Q8VGxsrm82mdevW+ey3LEvz5s1Tr1691KlTJ6WlpengwYPBmWwbkZOTo+uvv15dunRRz549lZGRobKyMp8xZ86cUWZmprp166bOnTtr9OjRKi8vD9KM24YlS5Zo+PDh3i8TczgcevPNN737WbPGPfvss7LZbHrkkUe821g3/5566inZbDaf26BBg7z7WTf/jh07pvvuu0/dunVTp06dNGzYMO3cudO7vzVeE4icVrJmzRplZWUpOztbpaWlSkhIUHp6uioqKoI9tTajqqpKCQkJWrx4sd/9zz//vHJzc5WXl6dt27bp8ssvV3p6us6cOXORZ9p2FBUVKTMzU1u3blVBQYFqa2t16623qqqqyjvm0Ucf1YYNG5Sfn6+ioiIdP35cd999dxBnHXx9+vTRs88+q5KSEu3cuVO33HKL7rrrLu3du1cSa9aYHTt26Le//a2GDx/us511q98111yjzz//3Hv7xz/+4d3Hup3vyy+/1KhRo9ShQwe9+eab2rdvnxYsWKArrrjCO6ZVXhMstIrk5GQrMzPT+3NdXZ0VGxtr5eTkBHFWbZcka+3atd6fPR6PZbfbrfnz53u3VVZWWhEREdYf//jHIMywbaqoqLAkWUVFRZZlnV2jDh06WPn5+d4x+/fvtyRZxcXFwZpmm3TFFVdYv/vd71izRpw6dcoaOHCgVVBQYN10003WjBkzLMvi31pDsrOzrYSEBL/7WDf/Hn/8cevGG2+sd39rvSbwTk4rqKmpUUlJidLS0rzbQkJClJaWpuLi4iDOrP04dOiQnE6nzxpGRUUpJSWFNfwOl8slSerataskqaSkRLW1tT7rNmjQIMXHx7Nu/1VXV6fVq1erqqpKDoeDNWtEZmam7rzzTp/1kfi31piDBw8qNjZWAwYM0Lhx43TkyBFJrFt91q9fr6SkJP3kJz9Rz549de2112rZsmXe/a31mkDktIITJ06orq5OMTExPttjYmLkdDqDNKv25dw6sYb183g8euSRRzRq1CgNHTpU0tl1Cw8PV3R0tM9Y1k3avXu3OnfurIiICD388MNau3athgwZwpo1YPXq1SotLVVOTs55+1i3+qWkpGj58uXatGmTlixZokOHDun73/++Tp06xbrV49///reWLFmigQMH6q233tKUKVM0ffp0vfrqq5Ja7zXhkvx/IQfao8zMTO3Zs8fnb/2o39VXX60PPvhALpdLr7/+uiZMmKCioqJgT6vNOnr0qGbMmKGCggJ17Ngx2NNpV26//Xbvfx8+fLhSUlLUt29f/elPf1KnTp2COLO2y+PxKCkpSb/5zW8kSddee6327NmjvLw8TZgwodUeh3dyWkH37t0VGhp63tXy5eXlstvtQZpV+3JunVhD/6ZOnaqNGzfqnXfeUZ8+fbzb7Xa7ampqVFlZ6TOedZPCw8P1ve99T4mJicrJyVFCQoIWLlzImtWjpKREFRUVuu666xQWFqawsDAVFRUpNzdXYWFhiomJYd2aKDo6WldddZU+/vhj/r3Vo1evXhoyZIjPtsGDB3v/zNdarwlETisIDw9XYmKiCgsLvds8Ho8KCwvlcDiCOLP2o3///rLb7T5r6Ha7tW3btkt6DS3L0tSpU7V27Vpt3rxZ/fv399mfmJioDh06+KxbWVmZjhw5ckmvmz8ej0fV1dWsWT1SU1O1e/duffDBB95bUlKSxo0b5/3vrFvTfPXVV/rkk0/Uq1cv/r3VY9SoUed9HcZHH32kvn37SmrF14QLuToa31q9erUVERFhLV++3Nq3b581efJkKzo62nI6ncGeWptx6tQpa9euXdauXbssSdYLL7xg7dq1yzp8+LBlWZb17LPPWtHR0dZf//pX68MPP7Tuuusuq3///tbXX38d5JkHz5QpU6yoqCjr3XfftT7//HPv7fTp094xDz/8sBUfH29t3rzZ2rlzp+VwOCyHwxHEWQffrFmzrKKiIuvQoUPWhx9+aM2aNcuy2WzW22+/bVkWa9ZU3/10lWWxbvWZOXOm9e6771qHDh2y/vnPf1ppaWlW9+7drYqKCsuyWDd/tm/fboWFhVnPPPOMdfDgQWvlypXWZZddZq1YscI7pjVeE4icVvTyyy9b8fHxVnh4uJWcnGxt3bo12FNqU9555x1L0nm3CRMmWJZ19iODc+fOtWJiYqyIiAgrNTXVKisrC+6kg8zfekmy/vCHP3jHfP3119bPf/5z64orrrAuu+wy68c//rH1+eefB2/SbcCDDz5o9e3b1woPD7d69OhhpaamegPHslizpvrfyGHd/BszZozVq1cvKzw83Ordu7c1ZswY6+OPP/buZ93827BhgzV06FArIiLCGjRokLV06VKf/a3xmmCzLMtq8ftNAAAAbRTX5AAAACMROQAAwEhEDgAAMBKRAwAAjETkAAAAIxE5AADASEQOAAAwEpEDAACMROQAAAAjETkAAMBIRA4AADASkQMAAIz0/wHr6O5aqg6ukAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if DATASET == \"ms\":\n",
    "    positive_label = \"relapsing-remitting multiple sclerosis\"\n",
    "    negative_label = \"normal\"\n",
    "else: \n",
    "    positive_label = \"NHBE\"\n",
    "    negative_label = \"A549\"\n",
    "\n",
    "X = np.array(single_feature)[:, 0]\n",
    "y = np.array(labels)\n",
    "\n",
    "pos = X[y == 1]\n",
    "neg = X[y == 0]\n",
    "\n",
    "plt.plot(pos, np.zeros_like(pos), 'x')\n",
    "plt.plot(neg, np.zeros_like(neg), 'o')\n",
    "plt.show()\n",
    "\n",
    "#TODO: Add 2D plots as well"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
